### PR TITLE
Fix version specification to include build # count

### PR DESCRIPTION
**Pull request by gdevenyi**
Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by gdevenyi** [comment]
I'm not sure about whether I should change the build number for this.

**Comment by izahn** [comment]
> I'm not sure about whether I should change the build number for this.

I think probably not, and in fact I don't think we need to rebuild at all. I think we can use https://conda-forge.org/docs/maintainer/infrastructure.html#skipping-ci-builds for that.

**Comment by izahn** [comment]
https://github.com/gdevenyi/rstudio-feedstock/blob/df1d6fb33007f4260e285da6334d6567f97893b2/recipe/build-desktop.sh#L8 also needs to be changed to extract `RSTUDIO_VERSION_SUFFIX` from the version string.

**Comment by gdevenyi** [comment]
> https://github.com/gdevenyi/rstudio-feedstock/blob/df1d6fb33007f4260e285da6334d6567f97893b2/recipe/build-desktop.sh#L8 also needs to be changed to extract `RSTUDIO_VERSION_SUFFIX` from the version string.

Good catch!

**Comment by izahn** [comment]
@conda-forge-admin please rerender

**Comment by izahn** [comment]
It seems that this PR may be more important then I realized, because the bot is having problems hashing the url template. The error at https://conda-forge.org/status/#version_updates says

```
The recipe did not change in the version migration, a URL did not hash, or there is jinja2 syntax the bot cannot handle!

Please check the URLs in your recipe with version '2022.02.2+485' to make sure they exist!

We also found the following errors:

 - could not hash URL template 'https://github.com/rstudio/rstudio/archive/refs/tags/v{{ version }}+485.tar.gz'
``` 

I bumped the build number, and I think we should merge it and see if that helps get the bot unstuck.

**Comment by izahn** [comment]
Any concerns with merging this @h-vetinari ? I'm hoping it will unstick https://conda-forge.org/status/#version_updates and get bot PRs flowing here.

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml]: Not sure what happens if we put a "+" in the version; we could replace it with a "." if we are fine with a 4-part version

**Review by h-vetinari — COMMENTED** [review]
Not sure if there are builds in CF that have a "+" in their version. It doesn't sound ideal to me, and even if it works, it's probably fragile / not well-supported.

Same open question if the bot will be smart enough to figure this out.

I'm kind of tempted to reuse the upstream build number for our build number (while allowing extra increments on our side; since upstream jumps between build numbers are usually big enough to squeeze in a few of our builds).

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:1]:
```suggestion
{% set version = "2022.02.2+485" %}
{% set pkg_version = version.split("+")[0] %}
{% set upstream_build_num = version.split("+")[1] %}
# reset the below to 0 when bumping upstream build num.
{% set cf_build_num = 1 %}
```

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml]:
```suggestion
  number: {{ upstream_build_num + cf_build_num }}
```

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml]:
```suggestion
  version: {{ pkg_version }}
```

**Review by h-vetinari — COMMENTED** [review]
Something like the below _might_ still be upgrade-bot-compatible.

**Comment by izahn** [comment]
> I'm kind of tempted to reuse the upstream build number for our build number (while allowing extra increments on our side; since upstream jumps between build numbers are usually big enough to squeeze in a few of our builds).

Update: Just noticed the proposed solution, let me think about it. ~~I like the idea in some ways, but the upstream build number is needed to construct the URL. If we increment it locally won't we have to track the upstream build number separately in order to construct that URL?~~

**Comment by h-vetinari** [comment]
> Update: Just noticed the proposed solution, let me think about it.

I updated my proposal because I had forgotten to get rid of the 4-part version.

**Comment by izahn** [comment]
Looking at error messages for other packages in https://conda-forge.org/status/#version_updates suggests to me that the "could not hash" message is a bit of a red herring. It looks like that error simply means that the code couldn't be downloaded from the constructed URL. 

For example, selenium currently has an error listed there that looks like this:
```
The recipe did not change in the version migration, a URL did not hash, or there is jinja2 syntax the bot cannot handle!

Please check the URLs in your recipe with version '4.1.5' to make sure they exist!

We also found the following errors:

 - could not hash URL template 'https://github.com/SeleniumHQ/selenium/archive/refs/tags/selenium-{{ version }}.tar.gz'
```
which is plain-vanilla in terms of the version template, but indeed https://github.com/SeleniumHQ/selenium/archive/refs/tags/selenium-4.1.5.tar.gz does not exist.

In our case I think what happens is that we currently set the version to `2022.02.2` and then manually append `+485` to the URL. The bot looks at the github repository and sees a version tag with `2022.02.2+485`. It tries to construct the URL and ends up with 
```
https://github.com/rstudio/rstudio/archive/refs/tags/v2022.02.2+485+485.tar.gz
```
which indeed doesn't exist -- it should be
```
https://github.com/rstudio/rstudio/archive/refs/tags/v2022.02.2+485.tar.gz
```
At the end of all this, I think @gdevenyi 's original version number fix should work. The `+` shouldn't cause any problems, since it is quoted and should therefore be treated by jinja as a string.

@h-vetinari please let me know if any of the above looks wrong to you.

**Comment by h-vetinari** [comment]
> The `+` shouldn't cause any problems, since it is quoted and should therefore be treated by jinja as a string.
> 
> @h-vetinari please let me know if any of the above looks wrong to you.

Yes, I believe that the bot most likely work either way, but I definitely prefer
```
rstudio-desktop    2022.02.2           h123456_485
```
to 
```
rstudio-desktop    2022.02.2+485       h123456_0
```

**Comment by izahn** [comment]
I'm not so sure, I think it's kind of confusing. For one thing the PR template says

>Reset the build number to 0 (if the version changed)

If you feel strongly about it I'm willing to go along, but my view is that just `{% set version = "2022.02.2+485" %}` as we currently have in the PR is both simpler and clearer.

**Comment by h-vetinari** [comment]
Having an upstream build number is a special case not covered by the template. Your first attempt was actually better in that regard, but now that you want the bot to take care of updates, you need to consume the upstream build number.

I disagree that the result is clearer or the implementation substantially simpler, but feel free to do whatever you want.

**Comment by izahn** [comment]
Totally agree that having upstream build number is atypical, but I think conda-forge has a pretty strong [convention that build numbers start at 0](https://conda-forge.org/docs/maintainer/updating_pkgs.html#updating-recipes). 

It's also not clear what upstream uses the build number for -- according to https://www.rstudio.com/blog/calendar-versioning-for-commercial-rstudio-products/ the third component of the version is the patch number, not a date. Given that I can't see what circumstance would lead upstream to cut a release that only incremented the build number and not the patch number. Basically I think it makes more sense to just treat the whole thing as the version number, and preserve the conventional meaning of the conda-forge build number.

**Comment by h-vetinari** [comment]
> but I think conda-forge has a pretty strong [convention that build numbers start at 0](https://conda-forge.org/docs/maintainer/updating_pkgs.html#updating-recipes).

I can think of several feedstocks off the top of my head where that isn't the case.

> It's also not clear what upstream uses the build number for [...]

IMO irrelevant. If we want to (the bot to) consume it, we should use it (but not as a "+xxx" in the version number; _maybe_ as a ".xxx").

In any case I'm not going to argue this point further.

**Comment by izahn** [comment]
I don't mean for this to be an argument, I'm trying to figure out what to do and it's not really clear to me what the best way to handle it is. I know the discussion is getting tedious, but I keep coming back to it because I'm unsure what to do, not because I want to argue about it.

**Comment by izahn** [comment]
Plus signs in the version number is uncommon but not absent in conda-forge:
```
$ mamba search -q "*=*+*"
Loading channels: ...working... done
# Name                       Version           Build  Channel             
gnuradio             3.8.1.0+boost170  py36h3f4a7ec_4  conda-forge         
gnuradio             3.8.1.0+boost170  py37h2884269_4  conda-forge         
gnuradio             3.8.1.0+boost170  py38h6dfa95d_4  conda-forge         
gnuradio-core        3.8.1.0+boost170  py36h176242b_4  conda-forge         
gnuradio-core        3.8.1.0+boost170  py37h929c215_4  conda-forge         
gnuradio-core        3.8.1.0+boost170  py38h444392f_4  conda-forge         
gnuradio-grc         3.8.1.0+boost170  py36hb479042_4  conda-forge         
gnuradio-grc         3.8.1.0+boost170  py37h737c77d_4  conda-forge         
gnuradio-grc         3.8.1.0+boost170  py38h8016156_4  conda-forge         
gnuradio-qtgui       3.8.1.0+boost170  py36h470b328_4  conda-forge         
gnuradio-qtgui       3.8.1.0+boost170  py37hdf6fba1_4  conda-forge         
gnuradio-qtgui       3.8.1.0+boost170  py38h0e66f8e_4  conda-forge         
gnuradio-uhd         3.8.1.0+boost170  py36hb479042_4  conda-forge         
gnuradio-uhd         3.8.1.0+boost170  py37h737c77d_4  conda-forge         
gnuradio-uhd         3.8.1.0+boost170  py38h8016156_4  conda-forge         
gnuradio-video-sdl   3.8.1.0+boost170  py36hb479042_4  conda-forge         
gnuradio-video-sdl   3.8.1.0+boost170  py37h737c77d_4  conda-forge         
gnuradio-video-sdl   3.8.1.0+boost170  py38h8016156_4  conda-forge         
gnuradio-zeromq      3.8.1.0+boost170  py36hbe23c60_4  conda-forge         
gnuradio-zeromq      3.8.1.0+boost170  py37h5bc3f6e_4  conda-forge         
gnuradio-zeromq      3.8.1.0+boost170  py38hf965e20_4  conda-forge         
postgis                  2.4.3+9.6.3               0  conda-forge         
postgis                  2.4.3+9.6.7               0  conda-forge         
postgis                   2.4.3+10.3               0  conda-forge         
purepng              0.2.0+0.py37fix py27h516909a_1001  conda-forge         
purepng              0.2.0+0.py37fix py36h516909a_1001  conda-forge         
purepng              0.2.0+0.py37fix py37h516909a_1001  conda-forge         
uhd                  3.15.0.0+boost170  py36h5fecffa_2  conda-forge         
uhd                  3.15.0.0+boost170  py36hce25f51_2  conda-forge         
uhd                  3.15.0.0+boost170  py37hb45bf65_2  conda-forge         
uhd                  3.15.0.0+boost170  py38ha84c3e3_2  conda-forge         
volk                  2.2.1+boost170  py36h831f99a_1  conda-forge         
volk                  2.2.1+boost170  py36h90709c0_1  conda-forge         
volk                  2.2.1+boost170  py37h3340039_1  conda-forge         
volk                  2.2.1+boost170  py38h950e882_1  conda-forge         
w3m                  0.5.3+git20220429      h2a70a0f_0  conda-forge  
```

**Comment by h-vetinari** [comment]
The search shows how _few_ packages do this (a homeopathic amount out of all conda feedstocks times _all_ builds for your platform); the gnuradio builds for example are a complete [exception](https://anaconda.org/conda-forge/gnuradio-core/files?page=32) (one build 2 years ago; page 32/34). More to the point (as you can see from the python build strings), no feedstock does this _currently_.

> but I keep coming back to it because I'm unsure what to do

I'm sure about what to do (modulo jinja mistakes), but I cannot seem to convince you, so I'll let you decide.

**Comment by izahn** [comment]
I'm totally convincible, if you're confident then lets do it. Can you layout the full solution, or just push directly to this PR branch, making sure that the constructed URL will be correct?

File: recipe/meta.yaml
Comment (izahn) [reply recipe/meta.yaml:1]: Ensuring that this is bot-update-compatible is the whole point of this PR, so let's make sure this proposal achieves that before committing to it.

Comment (izahn) [reply recipe/meta.yaml:1]: The `set pkg_version` line has mismatched quotes.

File: recipe/meta.yaml
Comment (izahn) [reply recipe/meta.yaml]: Do you know if the bot will reset this to zero, or will it leave this alone?

**Comment by gdevenyi** [comment]
> It's also not clear what upstream uses the build number for

Build number is used (at least) as the increment of their dailies for each release, https://dailies.rstudio.com/rstudio/spotted-wakerobin/desktop/jammy/

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:1]: Thanks, fixed

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml]: The bot will not touch this, which means that if we ever increment `cf_build_num`, we ~have to~ should decrement it again for the next version. I do not foresee this to be a great problem, because by default we're mostly going to package each upstream build only once (i.e. `cf_build_num` can stay at 0, untouched by the bot).

And really, the key part IMO is that the bot opens the PR at all. Cosmetic changes are easily possible with suggestions + commit (even on mobile).

**Comment by h-vetinari** [comment]
> > It's also not clear what upstream uses the build number for
> 
> Build number is used (at least) as the increment of their dailies for each release,

What's relevant for us and the bot are the published [tags](https://github.com/rstudio/rstudio/tags), which are a generously spaced selection of those dailies (hence why adding a small `cf_build_num` on top will not lead to collisions).

File: recipe/meta.yaml
Comment (izahn) [reply recipe/meta.yaml]: I looked around and based on the documentation in https://conda-forge.org/docs/maintainer/knowledge_base.html#including-a-no-mpi-build and the examples in https://github.com/conda-forge/hdf5-feedstock/pull/168/commits/91ec741b622b3df8203b20d0d233596d98cd3518#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aL4 and  https://github.com/conda-forge/libsharp-feedstock/pull/11/commits/0eacd9c5f4b0a73e40f49b0423fce9c787f7c79d#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aL4 it looks like the bot actually will do the right thing for us if we rename `cf_build_num` to just `build`. What do you think?

File: recipe/build-desktop.sh
Comment (h-vetinari) [inline recipe/build-desktop.sh]:
```suggestion
export RSTUDIO_VERSION_PATCH=$(echo ${PKG_VERSION} | cut -d. -f3 )
export RSTUDIO_VERSION_SUFFIX=+$(echo ${PKG_BUILDNUM})
```

**Comment by h-vetinari** [comment]
> I'm totally convincible, if you're confident then lets do it. Can you layout the full solution, or just push directly to this PR branch, making sure that the constructed URL will be correct?

Done. 🤞 I didn't overlook something. :)

**Comment by izahn** [comment]
> Done. crossed_fingers I didn't overlook something. :)

Great, thank you and sorry for the frustrating conversation yesterday.

One quick thing -- I looked around and based on the documentation in https://conda-forge.org/docs/maintainer/knowledge_base.html#including-a-no-mpi-build and the examples in https://github.com/conda-forge/hdf5-feedstock/commit/91ec741b622b3df8203b20d0d233596d98cd3518#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aL4 and https://github.com/conda-forge/libsharp-feedstock/commit/0eacd9c5f4b0a73e40f49b0423fce9c787f7c79d#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aL4 it looks like the bot actually will do the right thing for us if we rename `cf_build_num` to just `build`. What do you think?

**Comment by h-vetinari** [comment]
> it looks like the bot actually will do the right thing for us if we rename `cf_build_num` to just `build`. What do you think?

Good idea! I think the bot does handle certain jinja-variable names (IIRC `build_num` also works; not sure if documented).

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:5]: What was the benefit of moving `upstream_build_num`? now the grouping for `split("+")` is gone and my comment is not true anymore (or at least ambiguous).

Comment (izahn) [reply recipe/meta.yaml:5]: I moved it because I don't know exactly what rules the bot follows. In https://github.com/conda-forge/libsharp-feedstock/pull/11/commits/0eacd9c5f4b0a73e40f49b0423fce9c787f7c79d# it updates the first instance of `{{% set build ...` but not the second. I moved `set build` to guard against the possibility that the bot would be confused by `set upstream_build_num` and try to change that instead. All of that is totally speculative, I don't really know what the bot does outside of the examples I mentioned, but this seemed safer to me.

**Review by izahn — APPROVED** [review]
LGTM

File: recipe/meta.yaml
Comment (izahn) [reply recipe/meta.yaml:5]: This PR looks good to me, this is the only outstanding issue. Feel free to resolve it as you like, I'm fine if you don't think the bot will be confused by `upstream_build_number` and want to change it back. Once you are happy with some resolution here feel free to merge it!

Comment (h-vetinari) [reply recipe/meta.yaml:5]: As long as there was some reasoning behind it. :)

Comment (h-vetinari) [reply recipe/meta.yaml:5]: (I think it shouldn't affect the bot, but in any case, we can also sort down the rest of the split stuff after the bulld number with the next PR)

**Pull request merged**
